### ISSUE TITLE
Problem in desitarget.mtl with table ztargets

### ISSUE THREAD

**Issue by rncahn** [desihub/desitarget#83]
[problem.txt](https://github.com/desihub/desitarget/files/522546/problem.txt)

**Comment by sbailey** [comment]
Turning Bob's text file into actual text in the issue.  From Bob:

In the code below, taken from desitarget.mtl.py, the table targets is joined to the table zcat to make ztargets when zcat exists.  However, as you see below, there is something wrong with ztargets in that while its length is determined properly, the code comes to a halt when we ask for the maximum of ztargets["TAGETID']

```python
    n       = len(targets)
    targets = Table(targets)

    # FIXME (APC): NUMOBS vs. NUMOBS_MORE?

    # Create redshift catalog
    if zcat is not None:
        ztargets = join(targets, zcat['TARGETID', 'NUMOBS', 'Z', 'ZWARN'],
                            keys='TARGETID', join_type='outer')
        if ztargets.masked:
            unobs = ztargets['NUMOBS'].mask
            ztargets['NUMOBS'][unobs] = 0
        import IPython
        IPython.embed()
    else:
        ztargets           = targets.copy()
        ztargets['NUMOBS'] = np.zeros(n, dtype=np.int32)
        ztargets['Z']      = -1 * np.ones(n, dtype=np.float32)
        ztargets['ZWARN']  = -1 * np.ones(n, dtype=np.int32)

    ztargets['NUMOBS_MORE'] = np.maximum(0, calc_numobs(ztargets) - ztargets['NUMOBS'])
```

Results:
```
Epoch 1
5333 tiles to be included in fiberassign
Mon Oct 10 17:03:12 2016 Starting MTL
Python 3.5.2 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:53:06)
Type "copyright", "credits" or "license" for more information.

IPython 5.1.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: len(ztargets)
Out[1]: 48575429

In [2]: len(ztargets['TARGETID'])
Out[2]: 48575429

In [3]: ztargets['TARGETID'][0:10]
Out[3]:
<MaskedColumn name='TARGETID' dtype='int64' length=10>
    517888057
 105157091894
 183493163699
 299377321445
 332056090420
 660040370519
 755118763169
 909827825771
1031089724377
1090578292420

In [4]: max(targets['TARGETID'])
Out[4]: 4611685988046356629

In [5]: max(atargets['TARGETID'])  OOPS!  a typo
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
/global/homes/r/rncahn/desitarget/py/desitarget/mtl.py in <module>()
----> 1 max(atargets['TARGETID'])

NameError: name 'atargets' is not defined

In [6]: max(ztargets['TARGETID'])
```
Here are some more diagnostics, which might help:
```
---> 1 max(ztargets['RA'])

astropy/table/_column_mixins.pyx in astropy.table._column_mixins._MaskedColumnGetitemShim.__getitem__ (astropy/table/_column_mixins.c:1122)()

astropy/table/_column_mixins.pyx in astropy.table._column_mixins.base_getitem (astropy/table/_column_mixins.c:834)()

astropy/table/_column_mixins.pyx in astropy.table._column_mixins.masked_column_getitem (astropy/table/_column_mixins.c:1021)()

/global/common/edison/contrib/desi/conda/conda_3.5-20160913/lib/python3.5/site-packages/numpy/ma/core.py in __getitem__(self, indx)
   3115
   3116         """
-> 3117         dout = self.data[indx]
   3118         # We could directly use ndarray.__getitem__ on self.
   3119         # But then we would have to modify __array_finalize__ to prevent the

/global/common/edison/contrib/desi/conda/conda_3.5-20160913/lib/python3.5/site-packages/astropy/table/column.py in data(self)
    993     @property
    994     def data(self):
--> 995         out = self.view(ma.MaskedArray)
    996         # The following is necessary because of a bug in Numpy, which was
    997         # fixed in numpy/numpy#2703. The fix should be included in Numpy 1.8.0.

/global/common/edison/contrib/desi/conda/conda_3.5-20160913/lib/python3.5/site-packages/numpy/ma/core.py in view(self, dtype, type, fill_value)
   3042             try:
   3043                 if issubclass(dtype, ndarray):
-> 3044                     output = ndarray.view(self, dtype)
   3045                     dtype = None
   3046                 else:

/global/common/edison/contrib/desi/conda/conda_3.5-20160913/lib/python3.5/site-packages/numpy/ma/core.py in __array_finalize__(self, obj)
   2920         else:
   2921             _mask = nomask
-> 2922         self._mask = _mask
   2923         # Finalize the mask
   2924         if self._mask is not nomask:
```

I can avoid this problem by not asking for `max(ztargets['TARGETID'])`, however
I get the following warning:

> /global/common/edison/contrib/desi/conda/conda_3.5-20160913/lib/python3.5/site-packages/astropy/table/column.py:1095: MaskedArrayFutureWarning: setting an item on a masked array which has a shared mask will not copy the mask and also change the original mask array in the future.
Check the NumPy 1.11 release notes for more information.
  ma.MaskedArray.__setitem__(self, index, value)

So, somehow there is a problem with the table formed by the join, evidently related to the masking.

**Comment by rncahn** [comment]
I think I solved it. We just live with it
